### PR TITLE
Rework AsThough handling to allow choosing/affecting a specific alternate cast

### LINKED ISSUE #2087: Gisa and Geralf interaction with Risen Execution not implemented correctly
I confirmed the ruling and you should be given the option of choosing which effect you want, Risen Executioner's Tax or Gisa and Geralf's once a turn effect, when you try to cast Risen Executioner from the grave with Gisa and Geralf in play.

At the moment, with Gisa and Geralf in play and Risen Executioner in the grave, whenever you try to cast Risen Executioner from the grave, you are forced to cast it for the mana cost calculated by Risen Executioner's "tax" effect when cast from grave.

### LINKED ISSUE #8584: Vivien, Monster's Advocate takes top of deck Precedence to Bolas Citadel, not allowing for life-paid creature casting
Vivien, Monster's Advocate take top of deck Precedence to Bolas CItadel, not allowing for life-paid creature casting

### LINKED ISSUE #10519: One with the Multiverse
In this case I attacked with [[Narset, Enlightened Master]].  I played spells from exile for free ( [[Gilded Lotus]], [[One with the Multiverse]] ).  I then attempted to cast [[Promise of Loyalty]] off the top of the library.  It kept asking me to pay mana.

image

### LINKED ISSUE #10529: One with the Multiverse infinite free cast
Hello,

I was playing a local game against AI to test my Alrund Commander Deck and I played One with the Multiverse.
A turn after having casting it, I noticed that I was suddenly able to cast literally everything without the software asking me to pay for any of it. I played way more than one free spell and I didn't have an omniscience on the field. I don't know if this bug is already reported, but just in case.

Linked is a copy of the game log and a screenshot of the board state.

board state
[game log.txt](https://github.com/magefree/mage/files/11876726/game.log.txt)

### LINKED ISSUE #11058: One With the Multiverse doesn't let you cast cards for free from the top of your library

### PR DESCRIPTION
**Pull request by Susucre**
The main changes are the following:
- Now, it is possible for the player to choose between competing AsThough effects that allow to cast/play cards not the normal way (from non-hand, and/or with alternate cost). See notably the unit tests for One with the multiverse on that.
- There should be less side effects of alternate costs affecting other ways to cast. This is done with the help of MageIdentifier for both the AsThough and the alternate cost. There might still be the need to introduce more MageIdentifier, but I've done my best to find them.
- Again with the MageIdentifier improvements, it is now possible to differentiate multiple cast/play AsThough added by the same source. I do not think existing cards need that, but Court of Lochtwain do (it has a separate PR, not yet rebased)

I have added new tests and tweaked existing ones. So that helps with coverage of those effects. Also I have tested many of those cards live to ensure they were working as expected.

fix #2087 (Risen Executioner)
fix #8584 (Allow choosing between AsThough, case for Bolas's Citadel + Vivien, Mosnter's Advocate)
fix #11058 (One with the multiverse)
fix #10529 (One with the multiverse)
fix #10519 (One with the multiverse)

### CODE REVIEW

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java [hunk 1]
```diff
@@ -17,31 +19,46 @@ public interface ActivatedAbility extends Ability {
...
+            return new ActivationStatus(false, new HashSet<>());
         }
 
         /**
          * @param approvingObjectAbility ability that allows to activate/use current ability
          */
         public static ActivationStatus getTrue(Ability approvingObjectAbility, Game game) {
-            ApprovingObject approvingObject = approvingObjectAbility == null ? null : new ApprovingObject(approvingObjectAbility, game);
-            return new ActivationStatus(true, approvingObject);
+            Set<ApprovingObject> approvingObjects = new HashSet<>();
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/ActivatedAbility.java:45]: That code part looks bad:

1. Don't modify objects on read methods (except lazy init, but I don't see any needs in `null` data here at all -- just init field with a list).
2. No data duplication protections aka multiple calls of `getTrue` method. Fill an approving objects list and reading of the status must be diff methods, not in one.

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java [hunk 2]
```diff
@@ -180,13 +181,13 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
         // timing check
         //20091005 - 602.5d/602.5e
         boolean asInstant;
-        ApprovingObject approvingObject = game.getContinuousEffects()
+        Set<ApprovingObject> approvingObjects = game.getContinuousEffects()
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:184]: What about that comment?

shot_230905_211757

File: Mage/src/main/java/mage/abilities/SpellAbility.java [hunk 3]
```diff
@@ -146,7 +146,7 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
             if (getSpellAbilityType() == SpellAbilityType.BASE_ALTERNATE) {
                 Player player = game.getPlayer(playerId);
                 if (player != null
-                        && player.getCastSourceIdWithAlternateMana().contains(getSourceId())) {
+                        && player.getCastSourceIdWithAlternateMana().getOrDefault(getSourceId(), new HashSet<>()).contains(null)) {
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/SpellAbility.java:149]: `contains(null)` -- wtf here? Do not store any null objects in the list or map's keys.

File: Mage/src/main/java/mage/players/Player.java [hunk 4]
```diff
@@ -1054,13 +1051,28 @@ List<Integer> getMultiAmountWithIndividualConstraints(Outcome outcome, List<Mult
...
+     * without mana (null) or the mana set to manaCosts instead of its normal
+     * mana costs.
+     *
+     * @param sourceId  the source that can be cast without mana
+     * @param manaCosts alternate ManaCost, null if it can be cast without mana
+     *                  cost
+     * @param costs     alternate other costs you need to pay
+     * @param identifier if non-null, only apply the alternate mana when ApprovingSource if of that kind.
+     */
+    void setCastSourceIdWithAlternateMana(UUID sourceId, ManaCosts<ManaCost> manaCosts, Costs<Cost> costs, MageIdentifier identifier);
```
Comment (JayDi85) [inline Mage/src/main/java/mage/players/Player.java:1069]: Must be instruction how to use and add `MageIdentifier` for the cards and hows it works (like that: if you want to restrict alternative cast then use mage identifier like that, see example with card_name).

**Comment by JayDi85** [comment]
Also needs unit tests for new choose logic.

**Comment by Susucre** [comment]
More unit tests are in https://github.com/magefree/mage/pull/10973, they share the same base commit.

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java (hunk 1 above)
Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java:45]: BTW, if you used a set to protect from a duplication (`Set<ApprovingObject>`) then it will not work until override `equals` and `hashCode` (`ApprovingObject` don't have it).

Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java:45]: 3. Typo here. You are introduce and modify local variable `Set<ApprovingObject> approvingObjects`, but due code logic you must modify ability's `this.approvingObjects` here. Of course, if I've got the logic right (ability contains ApprovingObject's list and collect/keep it from a diff calls of `ActivationStatus.getTrue` on activation checks). Same mess with other `getTrue` methods. Must use diff names for class fields and local variables.
shot_230906_081807

File: Mage/src/main/java/mage/players/Player.java (hunk 4 above)
Comment (Susucre) [reply Mage/src/main/java/mage/players/Player.java:1069]: I did add comments on that in MageIdentifier class.

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java (hunk 2 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:184]: I have no idea and did not touch the `activatorId` current logic, this will require another investigation.

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java (hunk 1 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java]: Should be cleaner. Although `withoutApprovingObject` is weird and probably need a followup investigation. I did preserve the old behavior of 'there is no approving object, but the ActivationStatus.canActivate returns true.'

**Comment by Susucre** [comment]
I did edit the main post to provide explanation on what the PR is about. It should be good to review now.

File: Mage/src/main/java/mage/MageIdentifier.java [hunk 5]
```diff
@@ -19,7 +31,65 @@ public enum MageIdentifier {
     WishWatcher,
     GlimpseTheCosmosWatcher,
     SerraParagonWatcher,
-    OneWithTheMultiverseWatcher,
+    OneWithTheMultiverseWatcher("Without paying manacost"),
     JohannApprenticeSorcererWatcher,
-    KaghaShadowArchdruidWatcher
+    KaghaShadowArchdruidWatcher,
+    CourtOfLocthwainWatcher("Without paying manacost"),
```
Comment (xenohedron) [inline Mage/src/main/java/mage/MageIdentifier.java:37]: not part of this PR (will be used in the next one), but presumably still safe to include here anyway

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java [hunk 6]
```diff
@@ -7,41 +7,51 @@
...
+        private final Set<ApprovingObject> approvingObjects;
 
-        public ActivationStatus(boolean canActivate, ApprovingObject approvingObject) {
-            this.canActivate = canActivate;
-            this.approvingObject = approvingObject;
+        // If true, the Activation Status will not check if there is an approvingObject.
+        private final boolean forcedCanActivate;
+
+        public ActivationStatus(ApprovingObject approvingObject) {
+            this(new HashSet<>(Arrays.asList(approvingObject)));
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbility.java:28]: is this the best pattern to initialize for a single object? seems like an extra layer in there somewhere... maybe doesn't really matter though

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java [hunk 7]
```diff
@@ -7,41 +7,51 @@
...
 
-        public ApprovingObject getApprovingObject() {
-            return approvingObject;
+        public Set<ApprovingObject> getApprovingObjects() {
+            return approvingObjects;
         }
 
         public static ActivationStatus getFalse() {
-            return new ActivationStatus(false, null);
+            return new ActivationStatus(new HashSet<>());
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbility.java:50]: again, this seems like it could be an unnecessarily expensive way to initialize. Collections.emptySet() maybe? If never needs to be mutated there might be a way to use a static instance even

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java [hunk 8]
```diff
@@ -180,13 +181,13 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
...
                 .asThough(sourceId,
                         AsThoughEffectType.ACTIVATE_AS_INSTANT,
                         this,
                         controllerId,
                         game);
-        asInstant = approvingObject != null;
+        asInstant = !approvingObjects.isEmpty();
         asInstant |= (timing == TimingRule.INSTANT);
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:191]: I guess it was always this way, but weird to have this separated out instead of part of a single assignment

File: Mage/src/main/java/mage/abilities/SpellAbility.java [hunk 9]
```diff
@@ -141,12 +143,26 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
...
+            }
+
             // no mana restrict
             // Alternate spell abilities (Flashback, Overload) can't be cast with no mana to pay option
             if (getSpellAbilityType() == SpellAbilityType.BASE_ALTERNATE) {
                 Player player = game.getPlayer(playerId);
                 if (player != null
-                        && player.getCastSourceIdWithAlternateMana().contains(getSourceId())) {
+                        && player.getCastSourceIdWithAlternateMana()
+                                .getOrDefault(getSourceId(), new HashSet<>())
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/SpellAbility.java:163]: this is another example where Collections.emptySet() would make more sense than initializing an empty HashSet just to check what it contains

File: Mage/src/main/java/mage/abilities/common/PassAbility.java [hunk 10]
```diff
@@ -30,7 +31,7 @@ public PassAbility copy() {
 
     @Override
     public ActivationStatus canActivate(UUID playerId, Game game) {
-        return ActivationStatus.getTrue(this, game);
+        return new ActivationStatus(new ApprovingObject(this, game));
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/common/PassAbility.java:34]: can you explain this change?

File: Mage.Tests/src/test/java/org/mage/test/cards/single/bro/OneWithTheMultiverseTest.java [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+
+        assertPermanentCount(playerA, ogre, 0);
+        assertPermanentCount(playerA, piker, 2);
+        assertLibraryCount(playerA, ogre, 3);
+        assertHandCount(playerA, piker, 0);
+        assertTappedCount("Volcanic Island", true, 8 + 2);
+    }
+
+    @Test
+    public void castNormalFromHandThenFreeFromTop() {
```
Comment (xenohedron) [inline Mage.Tests/src/test/java/org/mage/test/cards/single/bro/OneWithTheMultiverseTest.java:336]: (for future, there's probably a way to parametrize tests like these)

File: Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java [hunk 12]
```diff
@@ -2243,7 +2238,7 @@ private boolean suppressAbilityPicker(ActivatedAbility ability, Game game) {
             }
 
             // hide on alternative cost activated
-            if (!getCastSourceIdWithAlternateMana().contains(ability.getSourceId())
+            if (!getCastSourceIdWithAlternateMana().getOrDefault(ability.getSourceId(), new HashSet<>()).contains(MageIdentifier.Default)
```
Comment (xenohedron) [inline Mage.Server.Plugins/Mage.Player.Human/src/mage/player/human/HumanPlayer.java:2241]: Use `Collections.emptySet()` instead of creating a `new HashSet<>()` when all you're going to do is check `contains()`.

File: Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java [hunk 13]
```diff
@@ -734,6 +734,7 @@ public void test_HostageTaker_CastFromExileAllParts() {
         checkExileCount("after exile 3", 3, PhaseStep.POSTCOMBAT_MAIN, playerA, "Curious Pair", 1);
         // play as adventure spell
         castSpell(3, PhaseStep.POSTCOMBAT_MAIN, playerA, "Treats to Share");
+        setChoice(playerA, "Hostage Taker"); // Not sure why there is an alternative there. No issue with using either. TODO: investigate?
```
Comment (xenohedron) [inline Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: any update on this?

**Review by xenohedron — COMMENTED** [review]
some questions and some potential performance improvements, but looks good overall

File: Mage.Tests/src/test/java/org/mage/test/cards/single/bro/OneWithTheMultiverseTest.java (hunk 11 above)
Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/cards/single/bro/OneWithTheMultiverseTest.java:336]: Difference between them was subtle enough I did not want to make them using shared methods. Sorry if that made the review work harder.

File: Mage/src/main/java/mage/abilities/common/PassAbility.java (hunk 10 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/common/PassAbility.java:34]: I'm not sure actually. I'll check.

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 14]
```diff
@@ -1080,25 +1099,34 @@ private void moveObjectToLibrary(UUID objectId, Ability source, Game game, boole
...
 
     @Override
-    public void setCastSourceIdWithAlternateMana(UUID sourceId, ManaCosts<ManaCost> manaCosts, Costs<Cost> costs) {
+    public void setCastSourceIdWithAlternateMana(UUID sourceId, ManaCosts<ManaCost> manaCosts, Costs<Cost> costs, MageIdentifier identifier) {
         // cost must be copied for data consistence between game simulations
-        castSourceIdWithAlternateMana.add(sourceId);
-        castSourceIdManaCosts.put(sourceId, manaCosts != null ? manaCosts.copy() : null);
-        castSourceIdCosts.put(sourceId, costs != null ? costs.copy() : null);
+        castSourceIdWithAlternateMana.computeIfAbsent(sourceId, k -> new HashSet<>());
+        castSourceIdWithAlternateMana.get(sourceId).add(identifier);
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage/src/main/java/mage/players/PlayerImpl.java:1105]: Because in my mind, `compute` function were not returning anything. Thanks for pointing it out, I'll simplify those.

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 15]
```diff
@@ -1311,6 +1366,68 @@ public boolean playLand(Card card, Game game, boolean ignoreTiming) {
...
+        public final ApprovingObjectResultStatus status;
+        public final ApprovingObject approvingObject; // not null iff status is CHOSEN
+
+        private ApprovingObjectResult(ApprovingObjectResultStatus status, ApprovingObject approvingObject) {
+            this.status = status;
+            this.approvingObject = approvingObject;
+        }
+    }
+
+    private ApprovingObjectResult chooseApprovingObject(Game game, List<ApprovingObject> possibleApprovingObjects, boolean required) {
```
[the remark this answers is not in the document]
Comment (Susucre) [reply Mage/src/main/java/mage/players/PlayerImpl.java:1385]: I think main changes are:
-> No longer distinguishing "consumable", as I did not find a single case where that made sense, all possible alternatives are considered and the player is asked to choose one of them (with the potential for no choice if required is false).
-> Now returns a ApprovingObjectResult, which includes an enum on how we got to that choice (relevant for required but no possible approving object to be different from not required and choice for none approving)
-> That's where we use the optional additional text for a MageIdentifier, to show up in the choice menu.

Aside from that yes that is doing somewhat the same as ContinuousEffect::asthough, the choice is just decoupled from returning the relevant approving objects, to be asked at the right time.

File: Mage/src/main/java/mage/MageIdentifier.java (hunk 5 above)
Comment (Susucre) [reply Mage/src/main/java/mage/MageIdentifier.java:37]: Yes, extra enum field not used is no harm.
Was easier to test the Court of Lochtwain code in addition to this branch. I can clear it if you mind it being there.

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java (hunk 6 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java:28]: If you have a better way to inline make a Set of one element, I'll take it. being called in a this does make it harder.

File: Mage/src/main/java/mage/abilities/common/PassAbility.java (hunk 10 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/common/PassAbility.java:34]: Alright after check analysis:
The reworked ActivationStatus is now always expecting to have an ApprovingObject when approved, for anything that might want to get additional info.
Having that ApprovingObject referencing the PassAbility does not alter anything, as PassAbilities are do not make use of complex activation logic.
image

File: Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java (hunk 13 above)
Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: So the choice the test think it has at that point are:
image

Treats to Share's Adventure should only be castable on its own from hand.

Testing the exact same actions than in the test, here is the first choice menu on casting the exiled by Hostage Taker adventure:
image
 
Clicking the first choice, second menu to choose approving object (like the test):
image
Both option lead to the adventure side starting being cast, with any color being accepted for the {G} payment.

Clicking the second choice, no menu to choose approving object, and start of cast of the creature side (with any type modification).

So really I'm not sure what is incorrect there for there to be two Approving objects. That's the only weird situation I've noticed among many manual tests (+ unit tests already existing or created in the PR)

Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: Oh actually that might just be Hostage Taker needing its own MageIdentifier. Will try to do that.

Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: Oh that goes deep in the "you may play|you may cast" with modifiers. If that's alright with you, I'll keep that investigation for the #10573 rework making use of the current AsThough changes to differentiate the modifications on how cards may be cast.

Comment (xenohedron) [reply Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: Yes, that's fine to leave it as is for this PR and track separately for future fix

Comment (Susucre) [reply Mage.Tests/src/test/java/org/mage/test/cards/cost/adventure/AdventureCardsTest.java:737]: made a note on that PR's big task list to remember.

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java (hunk 6 above)
Comment (xenohedron) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java:25]: Well, do you need to call a `this()`? I know it's generally good style for consistency, but it's not a complicated constructor. How about: 
```java
public ActivationStatus(ApprovingObject approvingObject) {
    this.forcedCanActivate = false;
    this.approvingObjects = Collections.singleton(approvingObject);
}
```

This is assuming you don't ever need to mutate the set (which I think is true but haven't confirmed).

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java [hunk 16]
```diff
@@ -180,15 +180,16 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
 
         // timing check
         //20091005 - 602.5d/602.5e
-        boolean asInstant;
-        Set<ApprovingObject> approvingObjects = game.getContinuousEffects()
+        Set<ApprovingObject> approvingObjects =game
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:183]: spacing

File: Mage/src/main/java/mage/players/PlayerImpl.java [hunk 17]
```diff
@@ -1219,8 +1222,12 @@ public boolean cast(SpellAbility originalAbility, Game game, boolean noMana, App
                         ? MageIdentifier.Default
                         : approvingObject.getApprovingAbility().getIdentifier();
 
-                if (!identifier.equals(MageIdentifier.Default)
-                        && getCastSourceIdWithAlternateMana().getOrDefault(ability.getSourceId(), new HashSet<>()).contains(identifier)) {
+                if(!getCastSourceIdWithAlternateMana().getOrDefault(ability.getSourceId(), Collections.emptySet()).contains(identifier)) {
```
Comment (xenohedron) [inline Mage/src/main/java/mage/players/PlayerImpl.java:1225]: did you change your IDE settings? prefer space between `if (` in the future

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java [hunk 18]
```diff
@@ -7,41 +7,50 @@
...
         }
 
-        /**
-         * @param approvingObjectAbility ability that allows to activate/use current ability
-         */
-        public static ActivationStatus getTrue(Ability approvingObjectAbility, Game game) {
-            ApprovingObject approvingObject = approvingObjectAbility == null ? null : new ApprovingObject(approvingObjectAbility, game);
-            return new ActivationStatus(true, approvingObject);
+        public static ActivationStatus withoutApprovingObject(boolean forcedCanActivate) {
+            return new ActivationStatus(forcedCanActivate, new HashSet<>());
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbility.java:53]: how about this one, is there a reason not to use `Collections.emptySet()`?

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java [hunk 19]
```diff
@@ -7,41 +7,50 @@
...
         }
 
         public boolean canActivate() {
-            return canActivate;
+            return forcedCanActivate || !approvingObjects.isEmpty();
         }
 
-        public ApprovingObject getApprovingObject() {
-            return approvingObject;
+        public Set<ApprovingObject> getApprovingObjects() {
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbility.java:42]: maybe add a comment here that the set is intended to be read-only?

**Comment by xenohedron** [comment]
A few more questions:
* Does this resolve #9512 ?
* [[Intet, the Dreamer]], [[Xanathar, Guild Kingpin]], and the Offering ability were among previously noted potential issues with asThough (side effects in applies method). Would your MageIdentifier be able to help with any of those? (Not required for this PR.)

File: Mage/src/main/java/mage/abilities/ActivatedAbility.java (hunk 6 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbility.java:25]: There should be no mutation after the ActivationStatus is created.

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java [hunk 20]
```diff
@@ -204,7 +206,13 @@ public ActivationStatus canActivate(UUID playerId, Game game) {
         //  game.inCheckPlayableState() can't be a help here cause some cards checking activating status,
         //  activatorId must be removed
         this.activatorId = playerId;
-        return new ActivationStatus(true, approvingObject);
+
+        if (approvingObjects.isEmpty()) {
+            return ActivationStatus.withoutApprovingObject(true);
+        }
+        else {
+            return new ActivationStatus(approvingObjects);
```
Comment (xenohedron) [inline Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:214]: Sorry, one more question- why no `true` parameter here anymore?

**Review by xenohedron — COMMENTED** [review]
Alright, looks good to me!

File: Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java (hunk 20 above)
Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:214]: Won't be able to look at the actual code for a day or two, but ActivationStatus really has three modes:

- Some approvers, it is the role of the caller if any of them is actually good enough (mana to activate, further restriction, etc...)
- Not approved (actually two way to construct that for convenience, the asFalse constructor, and the withoutApprover set with the false parameter)
- Approved, but without an explicit approver. I did not want to mess much with those cases as I did not understand all of it and was not the main focus point of the PR (being fixing a few bugs, and allow Court of Lochtwain to work without workaround).

Comment (xenohedron) [reply Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:214]: Sure, that's fine, I just noticed that the old line of code had the true parameter and the new code doesn't (in the else case with approving objects not empty), so I wanted to confirm that was intentional.

Comment (Susucre) [reply Mage/src/main/java/mage/abilities/ActivatedAbilityImpl.java:214]: It is intentional, and to be completely clear, the same behavior is preserved, but split in two cases:
If there is no approvingObject, the ActivationStatus is still forced to be approved.
If there is at least one approvingObject, the Approving Status being non-empty will expose that it is approved without the need to have a true parameter to force the status to approved.

**Review by xenohedron — APPROVED** [review]
